src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryDataLoader.kt

```diff
@@ -281,84 +301,124 @@ object ChocolateFactoryDataLoader {
+        } ?: run {
+            if (otherUpgradePattern.matches(itemName)) 0 else null
+        } ?: return
+
+        if (slotIndex == ChocolateFactoryAPI.timeTowerIndex) {
+            this.profileStorage?.timeTowerLevel = level
+        }
 
-                    ChocolateFactoryAPI.coachRabbitIndex -> ChocolateAmount.averageChocPerSecond(
-                        baseMultiplierIncrease = 0.01
-                    )
+        if (isMaxed) {
+            val otherUpgrade = ChocolateFactoryUpgrade(slotIndex, level, null)
+            list.add(otherUpgrade)
+            return
+        }
 
-                    else -> {
-                        val otherUpgrade = ChocolateFactoryUpgrade(slotIndex, level, upgradeCost)
-                        list.add(otherUpgrade)
-                        return
-                    }
-                }
+        val newAverageChocolate = when (slotIndex) {
+            ChocolateFactoryAPI.timeTowerIndex -> ChocolateAmount.averageChocPerSecond(includeTower = true)
+            ChocolateFactoryAPI.coachRabbitIndex -> ChocolateAmount.averageChocPerSecond(baseMultiplierIncrease = 0.01)
+            else -> {
+                val otherUpgrade = ChocolateFactoryUpgrade(slotIndex, level, upgradeCost)
+                list.add(otherUpgrade)
+                return
             }
         }
-        if (level == null || newAverageChocolate == null || upgradeCost == null) return
 
-        val extra = (newAverageChocolate - averageChocolate).round(2)
-        val effectiveCost = (upgradeCost / extra).round(2)
+        addUpgradeToList(list, slotIndex, level, upgradeCost, averageChocolate, newAverageChocolate, isRabbit = false)
+    }
 
+    private fun addUpgradeToList(
+        list: MutableList<ChocolateFactoryUpgrade>,
+        slotIndex: Int,
+        level: Int,
+        upgradeCost: Long?,
+        averageChocolate: Double,
+        newAverageChocolate: Double,
+        isRabbit: Boolean,
+    ) {
+        val extra = (newAverageChocolate - averageChocolate).round(2)
+        val effectiveCost = (upgradeCost!! / extra).round(2)
         val upgrade = ChocolateFactoryUpgrade(slotIndex, level, upgradeCost, extra, effectiveCost, isRabbit = isRabbit)
         list.add(upgrade)
     }
 
+    private fun handleRabbitWarnings(item: ItemStack, slotIndex: Int) {
+        val isGoldenRabbit = clickMeGoldenRabbitPattern.matches(item.name)
+        val warningConfig = config.rabbitWarning
+
+        if (clickMeRabbitPattern.matches(item.name) || isGoldenRabbit) {
+            if (config.rabbitWarning.rabbitWarning) {
+                SoundUtils.playBeepSound()
+            }
+
+            if (warningConfig.specialRabbitWarning
+                && (isGoldenRabbit || item.getSkullTexture() in specialRabbitTextures)
+            ) {
+                SoundUtils.repeatSound(
+                    100,
+                    warningConfig.repeatSound,
+                    SoundUtils.createSound(warningConfig.specialRabbitSound, 1f)
```

Comment:
Don't create a sound object each time. rather cache the sound object until the specialRabbitSound config value changes (use property)

This is not important. if we want to push this fix pr fast, we can simply add a todo comment in code